src/librustc_mir/transform/qualify_min_const_fn.rs

```diff
@@ -234,31 +238,47 @@ fn check_statement(
 }
 
 fn check_operand(
+    tcx: TyCtxt<'tcx>,
     operand: &Operand<'tcx>,
     span: Span,
+    def_id: DefId,
+    body: &Body<'tcx>
 ) -> McfResult {
     match operand {
         Operand::Move(place) | Operand::Copy(place) => {
-            check_place(place, span)
+            check_place(tcx, place, span, def_id, body)
         }
         Operand::Constant(_) => Ok(()),
     }
 }
 
 fn check_place(
+    tcx: TyCtxt<'tcx>,
     place: &Place<'tcx>,
     span: Span,
+    def_id: DefId,
+    body: &Body<'tcx>
 ) -> McfResult {
     place.iterate(|place_base, place_projection| {
         for proj in place_projection {
             match proj.elem {
                 ProjectionElem::Downcast(..) => {
                     return Err((span, "`match` or `if let` in `const fn` is unstable".into()));
                 }
+                ProjectionElem::Field(..) => {
+                    let base_ty = Place::ty_from(&place.base, &proj.base, body, tcx).ty;
+                    if let Some(def) = base_ty.ty_adt_def() {
+                        // No union field accesses in `const fn`
+                        if def.is_union() {
+                            if !feature_allowed(tcx, def_id, sym::const_fn_union) {
+                                return Err((span, "accessing union fields is unstable".into()));
```

Comment:
Don't we already check for this elsewhere? I expected the check to move, but why is there a new one?

Reply:
Oh I see, this is explained in https://github.com/rust-lang/rustc-guide/pull/424.